File: packages/emitter/testcases/functions/type-guards-v2/TypeGuardsV2.ts
export interface User {
  kind: "user";
  username: string;
  email: string;
}

export interface Admin {
  kind: "admin";
  adminId: number;
  permissions: string[];
}

export type Account = User | Admin;

// Type guard functions
export function isUser(account: Account): account is User {
  return account.kind === "user";
}

export function isAdmin(account: Account): account is Admin {
  return account.kind === "admin";
}

// Case 1: Negated guard - narrow the else branch
export function handleNotUser(account: Account): string {
  if (!isUser(account)) {
    // account is Admin here
    return `Admin ${account.adminId}`;
  } else {
    // account is User here (narrowed)
    return `User ${account.username}`;
  }
}

// Case 2: && guard - compound condition with narrowing
export function getUserWithValidEmail(account: Account): string {
  if (isUser(account) && account.email.length > 0) {
    return account.email;
  }
  return "no email";
}

// Case 3: && guard with method call
export function getUsernameUppercase(account: Account): string {
  if (isUser(account) && account.username !== "") {
    return account.username.toUpperCase();
  }
  return "anonymous";
}

File: packages/emitter/testcases/functions/type-guards-v2/TypeGuardsV2.cs
namespace TestCases.functions.typeguardsv2
{
    public class User
    {
        public string kind { get; set; }

        public string username { get; set; }

        public string email { get; set; }
    }
    public class Admin
    {
        public string kind { get; set; }

        public double adminId { get; set; }

        public global::System.Collections.Generic.List<string> permissions { get; set; }
    }

            public static class TypeGuardsV2
            {
                // type Account = global::Tsonic.Runtime.Union<User, Admin>

                public static bool isUser(Account account)
                    {
                    return account.kind == "user";
                    }

                public static bool isAdmin(Account account)
                    {
                    return account.kind == "admin";
                    }

                public static string handleNotUser(Account account)
                    {
                    if (!account.Is1())
                        {
                        return $"Admin {account.adminId}";
                        }
                    else
                    {
                        var account__1_1 = account.As1();
                        return $"User {account__1_1.username}";
                    }
                    }

                public static string getUserWithValidEmail(Account account)
                    {
                    if (account.Is1())
                    {
                        var account__1_1 = account.As1();
                        if (global::Tsonic.JSRuntime.String.length(account__1_1.email) > 0.0)
                            {
                            return account__1_1.email;
                            }
                    }
                        return "no email";
                    }

                public static string getUsernameUppercase(Account account)
                    {
                    if (account.Is1())
                    {
                        var account__1_1 = account.As1();
                        if (account__1_1.username != "")
                            {
                            return global::Tsonic.JSRuntime.String.toUpperCase(account__1_1.username);
                            }
                    }
                        return "anonymous";
                    }
            }
}